File: go/rng.go
// LombokAlgoritma — deterministic PRNGs: SplitMix64, xoshiro256++, PCG32 (SPEC §5)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
//
// These generators are NOT suitable for secrets (use crypto/rand).

package lombokalgoritma

import "math/bits"

// Default seeds (SPEC §5.2, §5.3).
const (
	DefaultXoshiroSeed    uint64 = 0x123456789abcdef0
	DefaultPcg32InitState uint64 = 0x853c49e6748fea9b
	DefaultPcg32InitSeq   uint64 = 0xda3e39cb94b95bdb
)

// SplitMix64 (Steele, Lea, Flood 2014) — used to expand a 64-bit seed into generator state.
type SplitMix64 struct {
	x uint64
}

// NewSplitMix64 returns a SplitMix64 whose state is seed.
func NewSplitMix64(seed uint64) *SplitMix64 { return &SplitMix64{x: seed} }

// Next returns the next 64-bit output.
func (s *SplitMix64) Next() uint64 {
	s.x += 0x9e3779b97f4a7c15
	z := s.x
	z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9
	z = (z ^ (z >> 27)) * 0x94d049bb133111eb
	return z ^ (z >> 31)
}

// Xoshiro256pp is xoshiro256++ (Blackman & Vigna), state seeded with four SplitMix64 outputs.
type Xoshiro256pp struct {
	s [4]uint64
}

// NewXoshiro256pp seeds a generator from seed (see DefaultXoshiroSeed).
func NewXoshiro256pp(seed uint64) *Xoshiro256pp {
	sm := NewSplitMix64(seed)
	g := &Xoshiro256pp{}
	for i := range g.s {
		g.s[i] = sm.Next()
	}
	return g
}

// Next returns the next 64-bit output.
func (g *Xoshiro256pp) Next() uint64 {
	s := &g.s
	result := bits.RotateLeft64(s[0]+s[3], 23) + s[0]
	t := s[1] << 17
	s[2] ^= s[0]
	s[3] ^= s[1]
	s[1] ^= s[2]
	s[0] ^= s[3]
	s[2] ^= t
	s[3] = bits.RotateLeft64(s[3], 45)
	return result
}

// NextFloat returns a float in [0, 1): (Next() >> 11) / 2^53 (exact).
func (g *Xoshiro256pp) NextFloat() float64 {
	return float64(g.Next()>>11) / (1 << 53)
}

// NextInt returns an unbiased integer in [0, n) by rejection sampling, 1 ≤ n ≤ 2^53 − 1.
// Any other n yields an OUT_OF_RANGE error.
func (g *Xoshiro256pp) NextInt(n int64) (int64, error) {
	if n < 1 || n > maxSafeInt {
		return 0, newErr(CodeOutOfRange, "NextInt: n must be in [1, 2^53)")
	}
	un := uint64(n)
	threshold := (-un) % un // 2^64 mod n
	for {
		r := g.Next()
		if r >= threshold {
			return int64(r % un), nil
		}
	}
}

// Pcg32 is PCG-XSH-RR 64/32 as in pcg-c's pcg32_srandom_r (O'Neill 2014).
type Pcg32 struct {
	state uint64
	inc   uint64
}

// NewPcg32 seeds a generator with initState and initSeq (see the DefaultPcg32* constants).
func NewPcg32(initState, initSeq uint64) *Pcg32 {
	p := &Pcg32{inc: initSeq<<1 | 1}
	p.step()
	p.state += initState
	p.step()
	return p
}

func (p *Pcg32) step() { p.state = p.state*6364136223846793005 + p.inc }

// Next returns the next 32-bit output.
func (p *Pcg32) Next() uint32 {
	old := p.state
	p.step()
	xs := uint32(((old >> 18) ^ old) >> 27)
	rot := int(old >> 59)
	return bits.RotateLeft32(xs, -rot)
}

// NextBounded returns an unbiased integer in [0, bound) (pcg32_boundedrand_r); bound = 0 yields
// an OUT_OF_RANGE error.
func (p *Pcg32) NextBounded(bound uint32) (uint32, error) {
	if bound == 0 {
		return 0, newErr(CodeOutOfRange, "NextBounded: bound must be in [1, 2^32)")
	}
	threshold := -bound % bound
	for {
		r := p.Next()
		if r >= threshold {
			return r % bound, nil
		}
	}
}

// NextFloat returns Next() / 2^32, a float in [0, 1) with 32 bits of precision.
func (p *Pcg32) NextFloat() float64 {
	return float64(p.Next()) / (1 << 32)
}

// maxSafeInt is 2^53 − 1 (ECMAScript Number.MAX_SAFE_INTEGER).
const maxSafeInt = 1<<53 - 1

File: typescript/src/core/rng.ts
// LombokAlgoritma — Random Number Generators
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
// CSPRNG: Web Crypto `getRandomValues` (platform-provided).
// Non-crypto, deterministic: xoshiro256++ (seeded by SplitMix64) and PCG32 (PCG-XSH-RR 64/32).
// Deterministic generators are NOT suitable for secrets.

import { AlgoError, OutOfRangeError } from './errors.js';
const MASK64 = (1n << 64n) - 1n;

/** Cryptographically secure random bytes — uses the platform Web Crypto API. */
export function randomBytes(n: number): Uint8Array {
  const buf = new Uint8Array(n);
  const c = (globalThis as { crypto?: { getRandomValues(a: Uint8Array): Uint8Array } }).crypto;
  if (c === undefined) {
    throw new AlgoError(
      'UNSUPPORTED',
      'randomBytes: globalThis.crypto.getRandomValues is unavailable',
    );
  }
  // getRandomValues is limited to 65536 bytes per call
  for (let off = 0; off < n; off += 65536) {
    c.getRandomValues(buf.subarray(off, Math.min(n, off + 65536)));
  }
  return buf;
}

/** Cryptographically secure random 32-bit unsigned integer. */
export function randomU32(): number {
  const b = randomBytes(4);
  return new DataView(b.buffer).getUint32(0, false);
}

/**
 * SplitMix64 — used to expand a 64-bit seed into generator state
 * (Steele, Lea, Flood 2014; reference: prng.di.unimi.it/splitmix64.c).
 */
export class SplitMix64 {
  private x: bigint;
  constructor(seed: bigint) {
    this.x = BigInt.asUintN(64, seed);
  }
  next(): bigint {
    this.x = (this.x + 0x9e3779b97f4a7c15n) & MASK64;
    let z = this.x;
    z = ((z ^ (z >> 30n)) * 0xbf58476d1ce4e5b9n) & MASK64;
    z = ((z ^ (z >> 27n)) * 0x94d049bb133111ebn) & MASK64;
    return z ^ (z >> 31n);
  }
}

function rotl64(x: bigint, k: bigint): bigint {
  return ((x << k) | (x >> (64n - k))) & MASK64;
}

/**
 * xoshiro256++ — fast non-crypto PRNG (Blackman & Vigna, prng.di.unimi.it/xoshiro256plusplus.c).
 * Period 2^256 − 1. State is initialised with four successive SplitMix64 outputs of `seed`.
 */
export class Xoshiro256pp {
  private readonly s: BigUint64Array;

  constructor(seed = 0x123456789abcdef0n) {
    this.s = new BigUint64Array(4);
    const sm = new SplitMix64(seed);
    for (let i = 0; i < 4; i++) this.s[i] = sm.next();
  }

  /** Next 64-bit output as an unsigned bigint. */
  next(): bigint {
    const s = this.s;
    const s0 = s[0] ?? 0n;
    const s1 = s[1] ?? 0n;
    const s2 = s[2] ?? 0n;
    const s3 = s[3] ?? 0n;
    const result = (rotl64((s0 + s3) & MASK64, 23n) + s0) & MASK64;
    const t = (s1 << 17n) & MASK64;
    const n2 = s2 ^ s0;
    const n3 = s3 ^ s1;
    const n1 = s1 ^ n2;
    const n0 = s0 ^ n3;
    s[0] = n0;
    s[1] = n1;
    s[2] = n2 ^ t;
    s[3] = rotl64(n3, 45n);
    return result;
  }

  /** Random float in [0, 1): top 53 bits / 2^53. */
  nextFloat(): number {
    return Number(this.next() >> 11n) / 2 ** 53;
  }

  /** Unbiased random integer in [0, n) via rejection sampling (1 ≤ n ≤ 2^53 − 1). */
  nextInt(n: number): number {
    if (!Number.isInteger(n) || n < 1 || n > Number.MAX_SAFE_INTEGER) {
      throw new OutOfRangeError('nextInt: n must be an integer in [1, 2^53)');
    }
    const bn = BigInt(n);
    const threshold = (1n << 64n) % bn; // reject the low `threshold` values
    for (;;) {
      const r = this.next();
      if (r >= threshold) return Number(r % bn);
    }
  }
}

/**
 * PCG32 (PCG-XSH-RR, 64-bit state → 32-bit output), per the pcg-c reference
 * `pcg32_srandom_r(initstate, initseq)` (O'Neill 2014, pcg-random.org).
 */
export class Pcg32 {
  private state = 0n;
  private readonly inc: bigint;

  constructor(initState = 0x853c49e6748fea9bn, initSeq = 0xda3e39cb94b95bdbn) {
    this.inc = ((BigInt.asUintN(64, initSeq) << 1n) | 1n) & MASK64;
    this.step();
    this.state = (this.state + BigInt.asUintN(64, initState)) & MASK64;
    this.step();
  }

  private step(): void {
    this.state = (this.state * 6364136223846793005n + this.inc) & MASK64;
  }

  /** Next 32-bit unsigned output. */
  next(): number {
    const old = this.state;
    this.step();
    const xorshifted = Number((((old >> 18n) ^ old) >> 27n) & 0xffffffffn) >>> 0;
    const rot = Number(old >> 59n);
    return ((xorshifted >>> rot) | (xorshifted << ((32 - rot) & 31))) >>> 0;
  }

  /** Unbiased integer in [0, bound) — pcg32_boundedrand_r. */
  nextBounded(bound: number): number {
    if (!Number.isInteger(bound) || bound < 1 || bound > 0xffffffff) {
      throw new OutOfRangeError('nextBounded: bound must be an integer in [1, 2^32)');
    }
    const threshold = (0x100000000 - bound) % bound;
    for (;;) {
      const r = this.next();
      if (r >= threshold) return r % bound;
    }
  }

  /** Random float in [0, 1) with 32 bits of precision. */
  nextFloat(): number {
    return this.next() / 2 ** 32;
  }
}
